flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/restore/RocksDBFullRestoreOperation.java

```diff
@@ -0,0 +1,276 @@
+			keySerializerProvider,
+			instanceBasePath,
+			instanceRocksDBPath,
+			dbOptions,
+			columnOptions,
+			nativeMetricMonitor);
+	}
+
+	/**
+	 * Restores all key-groups data that is referenced by the passed state handles.
+	 *  @param keyedStateHandles List of all key groups state handles that shall be restored.
+	 * @param db
+	 */
+	@Override
+	public RocksDBRestoreResult restore(Collection<KeyedStateHandle> keyedStateHandles, RocksDB db)
+		throws IOException, StateMigrationException, RocksDBException {
+		if (db == null) {
+			openDB();
+		} else {
+			this.db = db;
+		}
+		for (KeyedStateHandle keyedStateHandle : keyedStateHandles) {
+			if (keyedStateHandle != null) {
+
+				if (!(keyedStateHandle instanceof KeyGroupsStateHandle)) {
+					throw new IllegalStateException("Unexpected state handle type, " +
+						"expected: " + KeyGroupsStateHandle.class +
+						", but found: " + keyedStateHandle.getClass());
+				}
+				this.currentKeyGroupsStateHandle = (KeyGroupsStateHandle) keyedStateHandle;
+				restoreKeyGroupsInStateHandle();
+			}
+		}
+		return new RocksDBRestoreResult(this.db, lastCompletedCheckpointId, backendUID,
+			restoredSstFiles, defaultColumnFamilyHandle);
+	}
+
+	/**
+	 * Restore one key groups state handle.
+	 */
+	private void restoreKeyGroupsInStateHandle()
+		throws IOException, StateMigrationException, RocksDBException {
+		try {
+			currentStateHandleInStream = currentKeyGroupsStateHandle.openInputStream();
+			cancelStreamRegistry.registerCloseable(currentStateHandleInStream);
+			currentStateHandleInView = new DataInputViewStreamWrapper(currentStateHandleInStream);
+			restoreKVStateMetaData();
+			restoreKVStateData();
+		} finally {
+			if (cancelStreamRegistry.unregisterCloseable(currentStateHandleInStream)) {
+				IOUtils.closeQuietly(currentStateHandleInStream);
+			}
+		}
+	}
+
+	/**
+	 * Restore the KV-state / ColumnFamily meta data for all key-groups referenced by the current state handle.
+	 */
+	private void restoreKVStateMetaData() throws IOException, StateMigrationException, RocksDBException {
+
+		// isSerializerPresenceRequired flag is set to false, since for the RocksDB state backend,
+		// deserialization of state happens lazily during runtime; we depend on the fact
+		// that the new serializer for states could be compatible, and therefore the restore can continue
+		// without old serializers required to be present.
+		KeyedBackendSerializationProxy<K> serializationProxy =
+			new KeyedBackendSerializationProxy<>(userCodeClassLoader);
+
+		serializationProxy.read(currentStateHandleInView);
+
+		if (!isKeySerializerCompatibilityChecked) {
+			// check for key serializer compatibility; this also reconfigures the
+			// key serializer to be compatible, if it is required and is possible
+			TypeSerializerSchemaCompatibility<K> keySerializerSchemaCompat =
+				keySerializerProvider.setPreviousSerializerSnapshotForRestoredState(serializationProxy.getKeySerializerSnapshot());
```

Comment:
I think this is somewhat duplicated code in `RocksDBFullRestoreOperation` and `RocksDBIncrementalRestoreOperation`. Maybe we can unify the restore of meta data (reading, serializer checks, registering states,...) into the common base class?

Reply:
Agreed but please allow me to leave this as the last task of this review round, after emptying `restore` of rocksdb backend and remove those many `if` to prevent double-restore.

Reply:
Sure, no problem :)